generic_server.cc

```diff
@@ -115,15 +113,15 @@ server::~server()
 }
 
 future<> server::stop() {
-    if (!_stopping) {
+    if (!_gate.is_closed()) {
         co_await shutdown();
     }
 
-    co_await _all_connections_stopped.get_future();
+    co_await std::move(_all_connections_stopped);
 }
 
 future<> server::shutdown() {
-    _stopping = true;
+    _all_connections_stopped = _gate.close();
```

Comment:
I think we better mark this function as `noexcept` now, otherwise close may hang forever
if we don't get to call `c.shutdown()` for all open connection.

Reply:
Can you please elaborate? I don't understand.

`seastar::gate::close()` is `noexcept`, so if `server::shutdown()` throws, then it must be after `seastar::gate::close()` returns.

If `server::shutdown()` throws, then we don't know what's going to happen in callers in general (`server::shutdown()` is `public`). `server::stop()` is an internal caller of `server::shutdown()`; if the latter throws, then `server::stop()` will not get as far as waiting for the future returned by `_gate.close()` -- through `_all_connections_stopped`; instead the exception will escape from `server::stop()`.

I don't understand "close may hang forever".

Reply:
`_gate.close()` returns a future that will be resolved once all connections are destroyed.
If we don't get to `c.shutdown()` because an exception is thrown this `future<>` will never resolve.
Currently nothing  is likely to throw, but e.g. abort_accept is not marked as `noexcept` either so theoretically, it might throw an exception.
`make_lw_shared<size_t>(0);` can theoretically throw bad_alloc as well (in that case we're in deep trouble and seastar tasks would probably fail to allocate as well, but it just goes to show other stuff could happen as well.
BTW, this allocation seems extraneous, and coroutinizing this function could save it.

Reply:
Understood, thanks; I'll look into those for v4.

Reply:
@bhalevy

> BTW, this allocation seems extraneous, and coroutinizing this function could save it.

I'm finding it difficult to coroutinize `shutdown()`, for two reasons:
- Within the `parallel_for_each`, there are two (nested) lambdas that each capture (by reference). [Capturing lambdas should not be coroutines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rcoro-capture), so I can't seem to turn these lambdas themselves into coroutines, only the "outer code".
- The chain of futures returned by `shutdown()` is terminated with `_listeners_stopped` (moved); the caller of `shutdown()`, i.e. `stop()`, will wait for that to resolve. As far as I can tell, if I were to turn `shutdown()` into a coroutine, then I'd have to finish up the function with `co_await std::move(_listeners_stopped)`. That seems to be moving the wait for `_listeners_stopped` from the callers of `shutdown()` -- for example, from `server::stop()` -- into `shutdown()`. Is that OK?

Reply:
Hm, regarding the second question anyway -- that seems quite the "standard" with coroutinization.

Reply:
Right, I think I managed to do the coroutinization (minus the lambda as explained above).

Reply:
done in v4